File: Assets/Scripts/APlus/AnswerExpression.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;

namespace APlus
{
    /// <summary>答案表达式的值：要么是整数，要么是一段文本。四选一只比较 Display。</summary>
    public struct AnswerValue
    {
        public bool IsNumber;
        public int Number;
        public string Text;

        public static AnswerValue Of(int n) { AnswerValue v = new AnswerValue(); v.IsNumber = true; v.Number = n; v.Text = ""; return v; }
        public static AnswerValue Of(string s) { AnswerValue v = new AnswerValue(); v.Text = s ?? ""; return v; }

        public string Display { get { return IsNumber ? Number.ToString(CultureInfo.InvariantCulture) : Text; } }
    }

    /// <summary>
    /// 动态答案求值时能读到的世界状态。同一轮里，场景把同一份上下文画出来，
    /// 卷子用它判分 —— 所以挂钟慢几分钟、红榜排第几，场景和答案永远一致。
    /// </summary>
    public sealed class AnswerContext
    {
        public int Seed;
        public int Year;
        public int Attempt;
        public int Deaths;
        public int RoundSize;
        /// <summary>q.fill.07：挂钟比广播慢几分钟（每轮随机 0–15）。</summary>
        public int ClockOffset;
        public readonly Dictionary<string, string> Strings = new Dictionary<string, string>(StringComparer.Ordinal);
        /// <summary>@key 取文本表里的「意思」。</summary>
        public Func<string, string> Text = delegate (string key) { return "[" + key + "]"; };

        public int IntVar(string name)
        {
            switch (name)
            {
                case "year": return Year;
                case "attempt": return Attempt;
                case "deaths": return Deaths;
                case "round_size": return RoundSize;
                case "clock_offset": return ClockOffset;
            }
            throw new FormatException("unknown variable: " + name);
        }

        public static AnswerContext FromRun(RunState run, int roundSize, StringTable table)
        {
            AnswerContext c = new AnswerContext();
            c.Seed = run.RoundSeed;
            c.Year = run.Year;
            c.Attempt = run.Attempt;
            c.Deaths = run.TotalDeaths;
            c.RoundSize = roundSize;
            c.ClockOffset = new DeterministicRng(unchecked(run.RoundSeed ^ 0x51ED270B)).Next(16);
            if (table != null) c.Text = table.Get;

            string header = run.HeaderShowsCandidateNumber
                ? run.CandidateNumber
                : Format(c.Text("paper.header.count_fmt"), run.Attempt.ToString(CultureInfo.InvariantCulture));
            c.Strings["header"] = header;
            c.Strings["candidate"] = run.CandidateNumber;
            c.Strings["roster_next"] = run.RosterNumber(0);
            return c;
        }

        public static string Format(string template, string arg)
        {
            return (template ?? "").Replace("{0}", arg ?? "");
        }
    }

    /// <summary>
    /// answer_chain.csv 里 answer_expr 与 distractors 中 `=` 项的求值器。语法（参数用 ; 分隔，避开 CSV 的逗号）：
    ///
    ///   expr    := unary (('+' | '-') unary)*
    ///   unary   := '-'? primary
    ///   primary := int | var | after(n) | rand(a;b) | oneof(e;e;…) | @key | @key(expr) | $name | 'text' | (expr)
    ///
    /// rand / oneof 用每题独立的确定性随机：同一轮同一题永远得到同一个值。
    /// </summary>
    public sealed class AnswerExpression
    {
        readonly string _s;
        readonly AnswerContext _ctx;
        readonly DeterministicRng _rng;
        int _i;

        /// <summary>oneof 没被选中的候选（给 distractors = rest 用）。</summary>
        public readonly List<AnswerValue> Alternatives = new List<AnswerValue>();

        AnswerExpression(string source, AnswerContext ctx, DeterministicRng rng)
        {
            _s = source ?? "";
            _ctx = ctx;
            _rng = rng;
        }

        public static AnswerValue Evaluate(string source, AnswerContext ctx, DeterministicRng rng, List<AnswerValue> alternatives)
        {
            AnswerExpression e = new AnswerExpression(source, ctx, rng);
            AnswerValue v = e.ParseExpr();
            e.SkipSpace();
            if (e._i != e._s.Length) throw e.Error("unexpected trailing input");
            if (alternatives != null) alternatives.AddRange(e.Alternatives);
            return v;
        }

        AnswerValue ParseExpr()
        {
            AnswerValue left = ParseUnary();
            while (true)
            {
                SkipSpace();
                if (_i >= _s.Length) return left;
                char op = _s[_i];
                if (op != '+' && op != '-') return left;
                _i++;
                AnswerValue right = ParseUnary();
                if (!left.IsNumber || !right.IsNumber) throw Error("arithmetic on text");
                left = AnswerValue.Of(op == '+' ? left.Number + right.Number : left.Number - right.Number);
            }
        }

        AnswerValue ParseUnary()
        {
            SkipSpace();
            if (Peek('-'))
            {
                _i++;
                AnswerValue v = ParsePrimary();
                if (!v.IsNumber) throw Error("negating text");
                return AnswerValue.Of(-v.Number);
            }
            return ParsePrimary();
        }

        AnswerValue ParsePrimary()
        {
            SkipSpace();
            if (_i >= _s.Length) throw Error("unexpected end");
            char c = _s[_i];

            if (c >= '0' && c <= '9')
            {
                int start = _i;
                while (_i < _s.Length && _s[_i] >= '0' && _s[_i] <= '9') _i++;
                return AnswerValue.Of(int.Parse(_s.Substring(start, _i - start), CultureInfo.InvariantCulture));
            }
            if (c == '(')
            {
                _i++;
                AnswerValue v = ParseExpr();
                Expect(')');
                return v;
            }
            if (c == '\'')
            {
                _i++;
                int end = _s.IndexOf('\'', _i);
                if (end < 0) throw Error("unterminated text");
                string text = _s.Substring(_i, end - _i);
                _i = end + 1;
                return AnswerValue.Of(text);
            }
            if (c == '$')
            {
                _i++;
                string name = ReadName(false);
                string value;
                if (!_ctx.Strings.TryGetValue(name, out value)) throw Error("unknown string variable " + name);
                return AnswerValue.Of(value);
            }
            if (c == '@')
            {
                _i++;
                string key = ReadName(true);
                string text = _ctx.Text(key);
                SkipSpace();
                if (Peek('('))
                {
                    _i++;
                    AnswerValue arg = ParseExpr();
                    Expect(')');
                    text = AnswerContext.Format(text, arg.Display);
                }
                return AnswerValue.Of(text);
            }

            string ident = ReadName(false);
            SkipSpace();
            if (!Peek('(')) return AnswerValue.Of(_ctx.IntVar(ident));
            _i++;
            List<AnswerValue> args = new List<AnswerValue>();
            SkipSpace();
            if (!Peek(')'))
            {
                while (true)
                {
                    args.Add(ParseExpr());
                    SkipSpace();
                    if (Peek(';')) { _i++; continue; }
                    break;
                }
            }
            Expect(')');
            return Call(ident, args);
        }

        AnswerValue Call(string name, List<AnswerValue> args)
        {
            if (name == "after")
            {
                RequireNumbers(name, args, 1);
                return AnswerValue.Of(_ctx.Attempt >= args[0].Number ? 1 : 0);
            }
            if (name == "rand")
            {
                RequireNumbers(name, args, 2);
                int lo = Math.Min(args[0].Number, args[1].Number);
                int hi = Math.Max(args[0].Number, args[1].Number);
                return AnswerValue.Of(lo + _rng.Next(hi - lo + 1));
            }
            if (name == "oneof")
            {
                if (args.Count < 2) throw Error("oneof needs at least 2 choices");
                int pick = _rng.Next(args.Count);
                for (int i = 0; i < args.Count; i++) if (i != pick) Alternatives.Add(args[i]);
                return args[pick];
            }
            throw Error("unknown function " + name);
        }

        void RequireNumbers(string name, List<AnswerValue> args, int count)
        {
            if (args.Count != count) throw Error(name + " expects " + count + " argument(s)");
            for (int i = 0; i < args.Count; i++) if (!args[i].IsNumber) throw Error(name + " expects numbers");
        }

        string ReadName(bool allowDots)
        {
            int start = _i;
            while (_i < _s.Length)
            {
                char c = _s[_i];
                bool ok = (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == '_' || (allowDots && c == '.');
                if (!ok) break;
                _i++;
            }
            if (_i == start) throw Error("expected a name");
            return _s.Substring(start, _i - start);
        }

        void SkipSpace() { while (_i < _s.Length && _s[_i] == ' ') _i++; }
        bool Peek(char c) { return _i < _s.Length && _s[_i] == c; }

        void Expect(char c)
        {
            SkipSpace();
            if (!Peek(c)) throw Error("expected '" + c + "'");
            _i++;
        }

        FormatException Error(string message)
        {
            return new FormatException(message + " at " + _i + " in: " + _s);
        }

        /// <summary>FNV-1a：跨运行时稳定（string.GetHashCode 在 .NET 与 Mono 上不一致）。</summary>
        public static int StableHash(string s)
        {
            uint h = 2166136261u;
            byte[] bytes = Encoding.UTF8.GetBytes(s ?? "");
            for (int i = 0; i < bytes.Length; i++) { h ^= bytes[i]; h *= 16777619u; }
            return unchecked((int)h);
        }
    }
}

File: src/core/answerExpression.ts
import { DeterministicRng } from './rng';
import type { RunState } from './runState';
import type { StringTable } from './stringTable';

/** 答案表达式的值：要么是整数，要么是一段文本。四选一只比较 display。 */
export interface AnswerValue {
  isNumber: boolean;
  number: number;
  text: string;
}

export function numberValue(n: number): AnswerValue {
  return { isNumber: true, number: n, text: '' };
}

export function textValue(s: string): AnswerValue {
  return { isNumber: false, number: 0, text: s ?? '' };
}

export function display(v: AnswerValue): string {
  return v.isNumber ? String(v.number) : v.text;
}

export class FormatError extends Error {
  override name = 'FormatError';
}

export function formatTemplate(template: string, arg: string): string {
  return (template ?? '').split('{0}').join(arg ?? '');
}

/**
 * 动态答案求值时能读到的世界状态。同一轮里，场景把同一份上下文画出来，
 * 卷子用它判分 —— 所以挂钟慢几分钟、红榜排第几，场景和答案永远一致。
 */
export class AnswerContext {
  seed = 0;
  year = 0;
  attempt = 0;
  deaths = 0;
  roundSize = 0;
  /** q.fill.07：挂钟比广播慢几分钟（每轮随机 0–15）。 */
  clockOffset = 0;
  readonly strings = new Map<string, string>();
  /** @key 取文本表里的「意思」。 */
  text: (key: string) => string = (key) => '[' + key + ']';

  intVar(name: string): number {
    switch (name) {
      case 'year': return this.year;
      case 'attempt': return this.attempt;
      case 'deaths': return this.deaths;
      case 'round_size': return this.roundSize;
      case 'clock_offset': return this.clockOffset;
    }
    throw new FormatError('unknown variable: ' + name);
  }

  static fromRun(run: RunState, roundSize: number, table: StringTable | null): AnswerContext {
    const c = new AnswerContext();
    c.seed = run.roundSeed;
    c.year = run.year;
    c.attempt = run.attempt;
    c.deaths = run.totalDeaths;
    c.roundSize = roundSize;
    c.clockOffset = new DeterministicRng(run.roundSeed ^ 0x51ed270b).next(16);
    if (table) c.text = (key) => table.get(key);

    const header = run.headerShowsCandidateNumber
      ? run.candidateNumber
      : formatTemplate(c.text('paper.header.count_fmt'), String(run.attempt));
    c.strings.set('header', header);
    c.strings.set('candidate', run.candidateNumber);
    c.strings.set('roster_next', run.rosterNumber(0));
    return c;
  }
}

/**
 * answer_chain.csv 里 answer_expr 与 distractors 中 `=` 项的求值器。语法（参数用 ; 分隔，避开 CSV 的逗号）：
 *
 *   expr    := unary (('+' | '-') unary)*
 *   unary   := '-'? primary
 *   primary := int | var | after(n) | rand(a;b) | oneof(e;e;…) | @key | @key(expr) | $name | 'text' | (expr)
 *
 * rand / oneof 用每题独立的确定性随机：同一轮同一题永远得到同一个值。
 */
export class AnswerExpression {
  private i = 0;
  /** oneof 没被选中的候选（给 distractors = rest 用）。 */
  readonly alternatives: AnswerValue[] = [];

  private constructor(
    private readonly s: string,
    private readonly ctx: AnswerContext,
    private readonly rng: DeterministicRng,
  ) {}

  static evaluate(source: string, ctx: AnswerContext, rng: DeterministicRng, alternatives: AnswerValue[] | null): AnswerValue {
    const e = new AnswerExpression(source ?? '', ctx, rng);
    const v = e.parseExpr();
    e.skipSpace();
    if (e.i !== e.s.length) throw e.error('unexpected trailing input');
    if (alternatives) alternatives.push(...e.alternatives);
    return v;
  }

  private parseExpr(): AnswerValue {
    let left = this.parseUnary();
    for (;;) {
      this.skipSpace();
      if (this.i >= this.s.length) return left;
      const op = this.s[this.i];
      if (op !== '+' && op !== '-') return left;
      this.i++;
      const right = this.parseUnary();
      if (!left.isNumber || !right.isNumber) throw this.error('arithmetic on text');
      left = numberValue(op === '+' ? (left.number + right.number) | 0 : (left.number - right.number) | 0);
    }
  }

  private parseUnary(): AnswerValue {
    this.skipSpace();
    if (this.peek('-')) {
      this.i++;
      const v = this.parsePrimary();
      if (!v.isNumber) throw this.error('negating text');
      return numberValue(-v.number | 0);
    }
    return this.parsePrimary();
  }

  private parsePrimary(): AnswerValue {
    this.skipSpace();
    if (this.i >= this.s.length) throw this.error('unexpected end');
    const c = this.s[this.i];

    if (c >= '0' && c <= '9') {
      const start = this.i;
      while (this.i < this.s.length && this.s[this.i] >= '0' && this.s[this.i] <= '9') this.i++;
      return numberValue(parseInt(this.s.substring(start, this.i), 10));
    }
    if (c === '(') {
      this.i++;
      const v = this.parseExpr();
      this.expect(')');
      return v;
    }
    if (c === "'") {
      this.i++;
      const end = this.s.indexOf("'", this.i);
      if (end < 0) throw this.error('unterminated text');
      const text = this.s.substring(this.i, end);
      this.i = end + 1;
      return textValue(text);
    }
    if (c === '$') {
      this.i++;
      const name = this.readName(false);
      const value = this.ctx.strings.get(name);
      if (value === undefined) throw this.error('unknown string variable ' + name);
      return textValue(value);
    }
    if (c === '@') {
      this.i++;
      const key = this.readName(true);
      let text = this.ctx.text(key);
      this.skipSpace();
      if (this.peek('(')) {
        this.i++;
        const arg = this.parseExpr();
        this.expect(')');
        text = formatTemplate(text, display(arg));
      }
      return textValue(text);
    }

    const ident = this.readName(false);
    this.skipSpace();
    if (!this.peek('(')) return numberValue(this.ctx.intVar(ident));
    this.i++;
    const args: AnswerValue[] = [];
    this.skipSpace();
    if (!this.peek(')')) {
      for (;;) {
        args.push(this.parseExpr());
        this.skipSpace();
        if (this.peek(';')) { this.i++; continue; }
        break;
      }
    }
    this.expect(')');
    return this.call(ident, args);
  }

  private call(name: string, args: AnswerValue[]): AnswerValue {
    if (name === 'after') {
      this.requireNumbers(name, args, 1);
      return numberValue(this.ctx.attempt >= args[0].number ? 1 : 0);
    }
    if (name === 'rand') {
      this.requireNumbers(name, args, 2);
      const lo = Math.min(args[0].number, args[1].number);
      const hi = Math.max(args[0].number, args[1].number);
      return numberValue(lo + this.rng.next(hi - lo + 1));
    }
    if (name === 'oneof') {
      if (args.length < 2) throw this.error('oneof needs at least 2 choices');
      const pick = this.rng.next(args.length);
      for (let k = 0; k < args.length; k++) if (k !== pick) this.alternatives.push(args[k]);
      return args[pick];
    }
    throw this.error('unknown function ' + name);
  }

  private requireNumbers(name: string, args: AnswerValue[], count: number): void {
    if (args.length !== count) throw this.error(name + ' expects ' + count + ' argument(s)');
    if (args.some((a) => !a.isNumber)) throw this.error(name + ' expects numbers');
  }

  private readName(allowDots: boolean): string {
    const start = this.i;
    while (this.i < this.s.length) {
      const c = this.s[this.i];
      const ok = (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c === '_' || (allowDots && c === '.');
      if (!ok) break;
      this.i++;
    }
    if (this.i === start) throw this.error('expected a name');
    return this.s.substring(start, this.i);
  }

  private skipSpace(): void {
    while (this.i < this.s.length && this.s[this.i] === ' ') this.i++;
  }

  private peek(c: string): boolean {
    return this.i < this.s.length && this.s[this.i] === c;
  }

  private expect(c: string): void {
    this.skipSpace();
    if (!this.peek(c)) throw this.error("expected '" + c + "'");
    this.i++;
  }

  private error(message: string): FormatError {
    return new FormatError(message + ' at ' + this.i + ' in: ' + this.s);
  }
}
